src/vroom_errors.h

```diff
@@ -83,15 +85,12 @@ class vroom_errors {
   void warn_for_errors() const {
     if (!have_warned_ && rows_.size() > 0) {
       have_warned_ = true;
-      static auto warn = Rf_findFun(
-          Rf_install("warn"),
-          Rf_findVarInFrame(R_NamespaceRegistry, Rf_install("rlang")));
-      cpp11::sexp warn_call = Rf_lang3(
-          warn,
-          Rf_mkString(
-              "One or more parsing issues, see `problems()` for details"),
-          Rf_mkString("vroom_parse_issue"));
-      Rf_eval(warn_call, R_EmptyEnv);
+      static auto cli_warn = cpp11::package("cli")["cli_warn"];
+      cpp11::strings bullets({
+        "w"_nm = "One or more parsing issues, call {.fun problems} on your data frame for details, e.g.:",
+        " "_nm = "dat <- vroom(...)",
+        " "_nm = "problems(dat)"});
+      cli_warn(bullets, "class"_nm = "vroom_parse_issue", ".envir"_nm = R_EmptyEnv);
```

Comment:
Upon reading more of the history of this function, it looks like there's a specific reason to _not_ use `cpp11::package`. But I still think we can switch to `cli::cli_warn()` (as opposed to `rlang::warn()`).

https://github.com/tidyverse/vroom/commit/984a3e5e37e124feacfec3d184dbeb02eb1145c4

Reply:
I can somewhat explain this. If you use `cpp11::package()` to call `warn()`, which involves the unwind protect, then if a user tries to catch the classed warning with `withCallingHandlers()` and tries to promote the warning to an error, then it will actually crash R.

I've set up a draft PR to demonstrate this: https://github.com/DavisVaughan/vroom/pull/1

Here is an example where the user might catch the classed warning and promote it to an error. This works with CRAN vroom correctly, as you see here:

``` r
library(vroom)

make_warning <- function() {
  x <- vroom(
    I("a,b\n1.0,x\n"), 
    delim = ",", 
    col_types = "dd"
  )
  
  # Warning happens at print time so force a print
  print(x)
}

# This fails hard if we unwind protect the warning (aborts RStudio)
# - Try to throw error after catching the warning
withCallingHandlers(
  expr = make_warning(),
  vroom_parse_issue = function(cnd) {
    rlang::abort("oh no")
  }
)
#> Error:
#> ! oh no
```

If you run this with the PR installed, then it will crash R. I don't quite understand why, but it seems to definitely be related to the unwind protect.

It seems to be only be related to error calls. I tried a few other variations where I called `print()` or `warn()` in the condition handler, and I also tried muffling the warning, and that all seemed to work fine even with the unwind protect. I have included those examples below (This was run with the PR installed).

<details>

``` r
library(vroom)

make_warning <- function() {
  x <- vroom(
    I("a,b\n1.0,x\n"), 
    delim = ",", 
    col_types = "dd"
  )
  
  # Warning happens at print time so force a print
  print(x)
}

# Works fine:
# - Print some stuff
# - Muffle the condition
withCallingHandlers(
  expr = make_warning(),
  vroom_parse_issue = function(cnd) {
    print("oh no")
    rlang::cnd_muffle(cnd)
  }
)
#> [1] "oh no"
#> # A tibble: 1 × 2
#>       a     b
#>   <dbl> <dbl>
#> 1     1    NA

# Works fine:
# - Print some stuff
# - Don't muffle the condition
withCallingHandlers(
  expr = make_warning(),
  vroom_parse_issue = function(cnd) {
    print("oh no")
  }
)
#> [1] "oh no"
#> Warning: One or more parsing issues, see `problems()` for details
#> # A tibble: 1 × 2
#>       a     b
#>   <dbl> <dbl>
#> 1     1    NA

# Works fine:
# - Throw a new warning
# - Muffle the condition
withCallingHandlers(
  expr = make_warning(),
  vroom_parse_issue = function(cnd) {
    rlang::warn("oh no")
    rlang::cnd_muffle(cnd)
  }
)
#> Warning: oh no
#> # A tibble: 1 × 2
#>       a     b
#>   <dbl> <dbl>
#> 1     1    NA

# Works fine:
# - Throw a new warning
# - Don't muffle the condition
withCallingHandlers(
  expr = make_warning(),
  vroom_parse_issue = function(cnd) {
    rlang::warn("oh no")
  }
)
#> Warning: oh no
#> One or more parsing issues, see `problems()` for details
#> # A tibble: 1 × 2
#>       a     b
#>   <dbl> <dbl>
#> 1     1    NA
```

<sup>Created on 2022-05-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

</details>

If it were me, this would prompt me to include a test for `test_that("can promote the vroom parse warning to an error")` that uses `withCallingHandlers()` like in my example above. That would ensure we don't accidentally screw that feature up in the future.

Reply:
Regarding the `static` in front of:

```cpp
static auto warn = Rf_findFun(
           Rf_install("warn"),
           Rf_findVarInFrame(R_NamespaceRegistry, Rf_install("rlang")));
```

That means that the initialization of `warn` is only run once. It may be somewhat expensive to look up `warn()` from the rlang namespace (probably not for this case), so this ensures that `warn` is only looked up once, and is stored until R is shut down.

https://stackoverflow.com/questions/5567529/what-makes-a-static-variable-initialize-only-once

Reply:
Thanks @DavisVaughan for the additional analysis.

I agree adding that test is a good idea.